File: Zig/trie_utils/src/trie.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Trie (Prefix Tree) implementation for efficient string storage and prefix operations
/// Supports insert, search, delete, autocomplete, and pattern matching
pub fn Trie(comptime max_word_size: usize) type {
    return struct {
        const Self = @This();

        /// Node in the Trie
        const Node = struct {
            children: [26]?*Node,
            is_end: bool,
            word_count: usize, // Number of complete words in this subtree
            child_count: usize, // Number of child nodes

            fn init(allocator: Allocator) Allocator.Error!*Node {
                const node = try allocator.create(Node);
                @memset(&node.children, null);
                node.is_end = false;
                node.word_count = 0;
                node.child_count = 0;
                return node;
            }

            fn deinit(self: *Node, allocator: Allocator) void {
                for (&self.children) |child_opt| {
                    if (child_opt) |child| {
                        child.deinit(allocator);
                    }
                }
                allocator.destroy(self);
            }
        };

        root: *Node,
        allocator: Allocator,
        size: usize, // Total number of words

        /// Initialize a new Trie
        pub fn init(allocator: Allocator) Allocator.Error!Self {
            const root = try Node.init(allocator);
            return .{
                .root = root,
                .allocator = allocator,
                .size = 0,
            };
        }

        /// Free all allocated memory
        pub fn deinit(self: *Self) void {
            self.root.deinit(self.allocator);
            self.* = undefined;
        }

        /// Get the number of words in the trie
        pub fn count(self: Self) usize {
            return self.size;
        }

        /// Check if the trie is empty
        pub fn isEmpty(self: Self) bool {
            return self.size == 0;
        }

        /// Convert a character to an index (0-25)
        fn charToIndex(c: u8) ?usize {
            if (c >= 'a' and c <= 'z') {
                return @as(usize, c - 'a');
            } else if (c >= 'A' and c <= 'Z') {
                return @as(usize, c - 'A');
            }
            return null;
        }

        /// Convert an index back to a character
        fn indexToChar(index: usize) u8 {
            return @as(u8, @intCast(index + @as(usize, 'a')));
        }

        /// Insert a word into the trie
        /// Returns true if the word was newly inserted, false if it already existed or is invalid
        pub fn insert(self: *Self, word: []const u8) Allocator.Error!bool {
            if (word.len == 0 or word.len > max_word_size) return false;

            var current = self.root;
            var path_nodes: [max_word_size]*Node = undefined;
            var path_len: usize = 0;

            for (word) |c| {
                const idx = charToIndex(c) orelse return false;
                if (current.children[idx] == null) {
                    current.children[idx] = try Node.init(self.allocator);
                    current.child_count += 1;
                }
                path_nodes[path_len] = current;
                path_len += 1;
                current = current.children[idx].?;
            }

            if (current.is_end) {
                return false; // Word already exists
            }

            current.is_end = true;
            self.size += 1;

            // Update word_count along the path (including the final node)
            for (0..path_len) |i| {
                path_nodes[i].word_count += 1;
            }
            current.word_count += 1; // Also increment the final node

            return true;
        }

        /// Insert multiple words into the trie
        /// Returns the number of words actually inserted
        pub fn insertBatch(self: *Self, words: []const []const u8) Allocator.Error!usize {
            var inserted: usize = 0;
            for (words) |word| {
                if (try self.insert(word)) {
                    inserted += 1;
                }
            }
            return inserted;
        }

        /// Search for a word in the trie
        pub fn search(self: Self, word: []const u8) bool {
            if (word.len == 0) return false;

            var current = self.root;
            for (word) |c| {
                const idx = charToIndex(c) orelse return false;
                current = current.children[idx] orelse return false;
            }
            return current.is_end;
        }

        /// Check if any word starts with the given prefix
        pub fn startsWith(self: Self, prefix: []const u8) bool {
            if (prefix.len == 0) return self.size > 0;

            var current = self.root;
            for (prefix) |c| {
                const idx = charToIndex(c) orelse return false;
                current = current.children[idx] orelse return false;
            }
            return true;
        }

        /// Count words with a given prefix
        pub fn countPrefix(self: Self, prefix: []const u8) usize {
            if (prefix.len == 0) return self.size;

            var current = self.root;
            for (prefix) |c| {
                const idx = charToIndex(c) orelse return 0;
                current = current.children[idx] orelse return 0;
            }
            return current.word_count;
        }

        /// Get the node at the end of a prefix, or null if not found
        fn getNode(self: Self, prefix: []const u8) ?*Node {
            if (prefix.len == 0) return self.root;

            var current = self.root;
            for (prefix) |c| {
                const idx = charToIndex(c) orelse return null;
                current = current.children[idx] orelse return null;
            }
            return current;
        }

        /// Collect all words with a given prefix
        pub fn getWordsWithPrefix(self: Self, allocator: Allocator, prefix: []const u8, max_results: usize) Allocator.Error![][]u8 {
            const start_node = self.getNode(prefix) orelse return &[_][]u8{};

            var result = std.ArrayList([]u8).init(allocator);
            defer result.deinit();

            var buffer: [max_word_size]u8 = undefined;
            @memcpy(buffer[0..prefix.len], prefix);

            try self.collectWords(allocator, start_node, &buffer, prefix.len, &result, max_results);
            return result.toOwnedSlice();
        }

        /// Collect all words starting from a node
        fn collectWords(
            self: Self,
            allocator: Allocator,
            node: *Node,
            buffer: *[max_word_size]u8,
            depth: usize,
            result: *std.ArrayList([]u8),
            max_results: usize,
        ) Allocator.Error!void {
            if (result.items.len >= max_results) return;

            if (node.is_end) {
                const word = try allocator.dupe(u8, buffer[0..depth]);
                try result.append(word);
            }

            if (result.items.len >= max_results) return;

            for (0..26) |i| {
                if (node.children[i]) |child| {
                    buffer[depth] = indexToChar(i);
                    try self.collectWords(allocator, child, buffer, depth + 1, result, max_results);
                    if (result.items.len >= max_results) return;
                }
            }
        }

        /// Get all words in the trie
        pub fn getAllWords(self: Self, allocator: Allocator) Allocator.Error![][]u8 {
            return self.getWordsWithPrefix(allocator, "", self.size);
        }

        /// Delete a word from the trie
        /// Returns true if the word was found and deleted, false otherwise
        pub fn delete(self: *Self, word: []const u8) bool {
            if (word.len == 0) return false;
            const existed = self.search(word);
            if (!existed) return false; // Word doesn't exist
            self.deleteHelper(self.root, word, 0);
            return true;
        }

        /// Helper function for deletion - updates internal state
        fn deleteHelper(self: *Self, node: *Node, word: []const u8, depth: usize) void {
            if (depth == word.len) {
                node.is_end = false;
                self.size -= 1;
                node.word_count = if (node.word_count > 0) node.word_count - 1 else 0;
                return;
            }

            const c = word[depth];
            const idx = charToIndex(c) orelse return;
            const child = node.children[idx] orelse return;

            self.deleteHelper(child, word, depth + 1);

            // Update word_count for this node
            node.word_count = if (node.word_count > 0) node.word_count - 1 else 0;

            // Check if child should be deleted
            if (!child.is_end and child.child_count == 0) {
                child.deinit(self.allocator);
                node.children[idx] = null;
                node.child_count -= 1;
            }
        }

        /// Find the longest common prefix among all words
        /// Writes the prefix to the output buffer and returns the length
        pub fn longestCommonPrefix(self: Self, output: []u8) usize {
            if (self.size == 0) return 0;

            var len: usize = 0;
            var current = self.root;

            while (len < output.len) {
                // If this is an end node, stop
                if (current.is_end) break;

                // Find the single child
                var single_child_idx: ?usize = null;
                var child_count: usize = 0;

                for (0..26) |i| {
                    if (current.children[i] != null) {
                        child_count += 1;
                        single_child_idx = i;
                    }
                }

                if (child_count != 1) break;

                const idx = single_child_idx.?;
                output[len] = indexToChar(idx);
                len += 1;
                current = current.children[idx].?;
            }

            return len;
        }

        /// Pattern matching with wildcards
        /// '?' matches any single character
        /// '*' matches zero or more characters
        pub fn patternMatch(self: Self, allocator: Allocator, pattern: []const u8, max_results: usize) Allocator.Error![][]u8 {
            var result = std.ArrayList([]u8).init(allocator);
            defer result.deinit();

            var buffer: [max_word_size]u8 = undefined;
            try self.patternMatchHelper(self.root, pattern, 0, &buffer, 0, &result, max_results);

            return result.toOwnedSlice();
        }

        fn patternMatchHelper(
            self: Self,
            node: *Node,
            pattern: []const u8,
            pattern_idx: usize,
            buffer: *[max_word_size]u8,
            depth: usize,
            result: *std.ArrayList([]u8),
            max_results: usize,
        ) Allocator.Error!void {
            if (result.items.len >= max_results) return;

            // Handle wildcard '*'
            if (pattern_idx < pattern.len and pattern[pattern_idx] == '*') {
                // Zero or more characters
                // Try matching zero characters first
                try self.patternMatchHelper(node, pattern, pattern_idx + 1, buffer, depth, result, max_results);

                // Try matching one or more characters
                for (0..26) |i| {
                    if (node.children[i]) |child| {
                        buffer[depth] = indexToChar(i);
                        try self.patternMatchHelper(child, pattern, pattern_idx, buffer, depth + 1, result, max_results);
                    }
                }
                return;
            }

            // End of pattern
            if (pattern_idx == pattern.len) {
                if (node.is_end) {
                    const word = try self.allocator.dupe(u8, buffer[0..depth]);
                    try result.append(word);
                }
                return;
            }

            // Handle wildcard '?'
            if (pattern[pattern_idx] == '?') {
                for (0..26) |i| {
                    if (node.children[i]) |child| {
                        buffer[depth] = indexToChar(i);
                        try self.patternMatchHelper(child, pattern, pattern_idx + 1, buffer, depth + 1, result, max_results);
                    }
                }
                return;
            }

            // Exact character match
            const idx = charToIndex(pattern[pattern_idx]) orelse return;
            const child = node.children[idx] orelse return;
            buffer[depth] = pattern[pattern_idx];
            try self.patternMatchHelper(child, pattern, pattern_idx + 1, buffer, depth + 1, result, max_results);
        }

        /// Count total number of nodes in the trie
        pub fn nodeCount(self: Self) usize {
            return self.countNodes(self.root);
        }

        fn countNodes(self: Self, node: *Node) usize {
            var node_count: usize = 1;
            for (0..26) |i| {
                if (node.children[i]) |child| {
                    node_count += self.countNodes(child);
                }
            }
            return node_count;
        }

        /// Estimate memory usage in bytes
        pub fn memoryUsage(self: Self) usize {
            const nodes = self.nodeCount();
            // Each node: 26 pointers (208 bytes on 64-bit) + 2 bools/size_t
            return nodes * @sizeOf(Node);
        }

        /// Clear all words from the trie
        pub fn clear(self: *Self) Allocator.Error!void {
            // Deallocate all children of root
            for (0..26) |i| {
                if (self.root.children[i]) |child| {
                    child.deinit(self.allocator);
                    self.root.children[i] = null;
                }
            }
            self.root.child_count = 0;
            self.root.word_count = 0;
            self.root.is_end = false;
            self.size = 0;
        }
    };
}

// Tests
const testing = std.testing;
const TrieType = Trie(256);

test "Trie - basic insert and search" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    try testing.expect(try trie.insert("hello"));
    try testing.expect(try trie.insert("world"));
    try testing.expect(try trie.insert("hi"));

    try testing.expect(trie.search("hello"));
    try testing.expect(trie.search("world"));
    try testing.expect(trie.search("hi"));
    try testing.expect(!trie.search("hell"));
    try testing.expect(!trie.search("helloworld"));
}

test "Trie - duplicate insert" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    try testing.expect(try trie.insert("test"));
    try testing.expect(!try trie.insert("test")); // Second insert returns false
    try testing.expect(trie.count() == 1);
}

test "Trie - prefix operations" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("apple");
    _ = try trie.insert("app");
    _ = try trie.insert("application");
    _ = try trie.insert("apply");

    try testing.expect(trie.startsWith("app"));
    try testing.expect(trie.startsWith("appl"));
    try testing.expect(!trie.startsWith("banana"));

    try testing.expect(trie.countPrefix("app") == 4);
    try testing.expect(trie.countPrefix("appl") == 3);
    try testing.expect(trie.countPrefix("apple") == 1);
    try testing.expect(trie.countPrefix("applex") == 0);
}

test "Trie - autocomplete" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("cat");
    _ = try trie.insert("car");
    _ = try trie.insert("card");
    _ = try trie.insert("care");
    _ = try trie.insert("careful");
    _ = try trie.insert("dog");

    const words = try trie.getWordsWithPrefix(allocator, "car", 10);
    defer {
        for (words) |w| allocator.free(w);
        allocator.free(words);
    }

    try testing.expect(words.len == 4);

    // Check that all words start with "car"
    for (words) |word| {
        try testing.expect(std.mem.startsWith(u8, word, "car"));
    }
}

test "Trie - delete" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("hello");
    _ = try trie.insert("hell");
    _ = try trie.insert("help");

    try testing.expect(trie.search("hello"));
    try testing.expect(trie.delete("hello"));
    try testing.expect(!trie.search("hello"));
    try testing.expect(trie.search("hell"));
    try testing.expect(trie.search("help"));
    try testing.expect(!trie.delete("hello")); // Already deleted
}

test "Trie - longest common prefix" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("flower");
    _ = try trie.insert("flow");
    _ = try trie.insert("flight");

    var buf: [256]u8 = undefined;
    const lcp_len = trie.longestCommonPrefix(&buf);
    try testing.expectEqualSlices(u8, "fl", buf[0..lcp_len]);
}

test "Trie - pattern match with ?" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("cat");
    _ = try trie.insert("car");
    _ = try trie.insert("card");
    _ = try trie.insert("care");
    _ = try trie.insert("bat");
    _ = try trie.insert("bar");

    const matches = try trie.patternMatch(allocator, "ca?", 10);
    defer {
        for (matches) |m| allocator.free(m);
        allocator.free(matches);
    }

    try testing.expect(matches.len == 2); // cat, car
}

test "Trie - pattern match with *" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("cat");
    _ = try trie.insert("car");
    _ = try trie.insert("card");

    const matches = try trie.patternMatch(allocator, "ca*", 10);
    defer {
        for (matches) |m| allocator.free(m);
        allocator.free(matches);
    }

    try testing.expect(matches.len == 3);
}

test "Trie - batch insert" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    const words = [_][]const u8{ "one", "two", "three", "four", "five" };
    const inserted = try trie.insertBatch(&words);

    try testing.expectEqual(@as(usize, 5), inserted);
    try testing.expectEqual(@as(usize, 5), trie.count());
}

test "Trie - empty trie" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    try testing.expect(trie.isEmpty());
    try testing.expectEqual(@as(usize, 0), trie.count());
    try testing.expect(!trie.search("anything"));
    try testing.expect(!trie.startsWith("any"));
}

test "Trie - clear" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("one");
    _ = try trie.insert("two");
    _ = try trie.insert("three");

    try testing.expectEqual(@as(usize, 3), trie.count());

    try trie.clear();

    try testing.expect(trie.isEmpty());
    try testing.expectEqual(@as(usize, 0), trie.count());
}

test "Trie - node count and memory usage" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("a");
    _ = try trie.insert("ab");
    _ = try trie.insert("abc");

    const nodes = trie.nodeCount();
    try testing.expect(nodes >= 3); // At least 3 nodes

    const mem = trie.memoryUsage();
    try testing.expect(mem > 0);
}

test "Trie - case insensitive" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("Hello");
    _ = try trie.insert("WORLD");

    try testing.expect(trie.search("hello"));
    try testing.expect(trie.search("HELLO"));
    try testing.expect(trie.search("world"));
    try testing.expect(trie.search("World"));
}

test "Trie - single character words" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    _ = try trie.insert("a");
    _ = try trie.insert("b");
    _ = try trie.insert("c");

    try testing.expect(trie.search("a"));
    try testing.expect(trie.search("b"));
    try testing.expect(trie.search("c"));
    try testing.expectEqual(@as(usize, 3), trie.count());
}

test "Trie - get all words" {
    const allocator = testing.allocator;
    var trie = try TrieType.init(allocator);
    defer trie.deinit();

    const test_words = [_][]const u8{ "apple", "banana", "cherry", "date" };
    for (test_words) |word| {
        _ = try trie.insert(word);
    }

    const all_words = try trie.getAllWords(allocator);
    defer {
        for (all_words) |w| allocator.free(w);
        allocator.free(all_words);
    }

    try testing.expectEqual(@as(usize, 4), all_words.len);
}

File: JavaScript/trie_utils/trie.js
/**
 * Trie (Prefix Tree) Utility
 * A tree-like data structure for efficient string retrieval and prefix-based operations.
 * 
 * Features:
 * - Insert, search, delete operations
 * - Prefix search / autocomplete
 * - Word frequency tracking
 * - Pattern matching with wildcards
 * 
 * Time Complexity:
 * - Insert: O(m) where m is the length of the word
 * - Search: O(m)
 * - Delete: O(m)
 * - Prefix search: O(m + n) where n is total characters in results
 * 
 * Space Complexity: O(n * m) where n is number of words, m is average length
 * 
 * Zero external dependencies - pure JavaScript implementation
 */

/**
 * Trie node class
 */
class TrieNode {
    constructor() {
        this.children = {};      // Map of character to TrieNode
        this.isEndOfWord = false; // Marks the end of a complete word
        this.frequency = 0;       // Word frequency counter
        this.data = null;         // Optional associated data
    }
}

/**
 * Trie data structure implementation
 */
class Trie {
    constructor() {
        this.root = new TrieNode();
        this.wordCount = 0;
    }

    /**
     * Insert a word into the trie
     * @param {string} word - The word to insert
     * @param {*} data - Optional data to associate with the word
     * @returns {Trie} Returns this for chaining
     */
    insert(word, data = null) {
        if (!word || typeof word !== 'string') {
            throw new Error('Word must be a non-empty string');
        }

        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
        }
        
        if (!node.isEndOfWord) {
            this.wordCount++;
        }
        
        node.isEndOfWord = true;
        node.frequency++;
        if (data !== null) {
            node.data = data;
        }
        
        return this;
    }

    /**
     * Search for a word in the trie
     * @param {string} word - The word to search for
     * @returns {boolean} True if the word exists in the trie
     */
    search(word) {
        if (!word || typeof word !== 'string') {
            return false;
        }

        const node = this._findNode(word);
        return node !== null && node.isEndOfWord;
    }

    /**
     * Check if any word in the trie starts with the given prefix
     * @param {string} prefix - The prefix to check
     * @returns {boolean} True if any word has this prefix
     */
    startsWith(prefix) {
        if (!prefix || typeof prefix !== 'string') {
            return false;
        }

        return this._findNode(prefix) !== null;
    }

    /**
     * Delete a word from the trie
     * @param {string} word - The word to delete
     * @returns {boolean} True if the word was deleted
     */
    delete(word) {
        if (!word || typeof word !== 'string') {
            return false;
        }

        const result = { deleted: false };
        this._deleteHelper(this.root, word, 0, result);
        
        if (result.deleted) {
            this.wordCount--;
        }
        
        return result.deleted;
    }

    /**
     * Recursive helper for delete operation
     * @private
     */
    _deleteHelper(node, word, index, result) {
        if (index === word.length) {
            if (node.isEndOfWord) {
                node.isEndOfWord = false;
                node.frequency = 0;
                node.data = null;
                result.deleted = true;
                return Object.keys(node.children).length === 0;
            }
            return false;
        }

        const char = word[index];
        const childNode = node.children[char];
        
        if (!childNode) {
            return false;
        }

        const shouldDeleteChild = this._deleteHelper(childNode, word, index + 1, result);

        if (shouldDeleteChild && !childNode.isEndOfWord) {
            delete node.children[char];
            return Object.keys(node.children).length === 0;
        }

        return false;
    }

    /**
     * Get all words with a given prefix (autocomplete)
     * @param {string} prefix - The prefix to search for
     * @param {number} limit - Maximum number of results (default: 10)
     * @returns {Array<{word: string, frequency: number, data: *}>}
     */
    autocomplete(prefix, limit = 10) {
        if (!prefix || typeof prefix !== 'string') {
            return [];
        }

        const node = this._findNode(prefix);
        if (!node) {
            return [];
        }

        const results = [];
        this._collectWords(node, prefix, results, limit);
        
        // Sort by frequency (descending)
        results.sort((a, b) => b.frequency - a.frequency);
        
        return results.slice(0, limit);
    }

    /**
     * Collect all words from a given node
     * @private
     */
    _collectWords(node, prefix, results, limit) {
        if (results.length >= limit) {
            return;
        }

        if (node.isEndOfWord) {
            results.push({
                word: prefix,
                frequency: node.frequency,
                data: node.data
            });
        }

        for (const char of Object.keys(node.children).sort()) {
            this._collectWords(node.children[char], prefix + char, results, limit);
        }
    }

    /**
     * Find the node at the end of a given prefix
     * @private
     */
    _findNode(prefix) {
        let node = this.root;
        for (const char of prefix) {
            if (!node.children[char]) {
                return null;
            }
            node = node.children[char];
        }
        return node;
    }

    /**
     * Get all words in the trie
     * @param {number} limit - Maximum number of results
     * @returns {Array<{word: string, frequency: number, data: *}>}
     */
    getAllWords(limit = Infinity) {
        const results = [];
        this._collectWords(this.root, '', results, limit);
        return results;
    }

    /**
     * Search for words matching a pattern with wildcards
     * @param {string} pattern - Pattern with '?' as wildcard (matches any single char)
     * @param {number} limit - Maximum number of results
     * @returns {Array<{word: string, frequency: number, data: *}>}
     */
    searchPattern(pattern, limit = 10) {
        if (!pattern || typeof pattern !== 'string') {
            return [];
        }

        const results = [];
        this._searchPatternHelper(this.root, pattern, 0, '', results, limit);
        results.sort((a, b) => b.frequency - a.frequency);
        return results.slice(0, limit);
    }

    /**
     * Recursive helper for pattern search
     * @private
     */
    _searchPatternHelper(node, pattern, index, currentWord, results, limit) {
        if (results.length >= limit) {
            return;
        }

        if (index === pattern.length) {
            if (node.isEndOfWord) {
                results.push({
                    word: currentWord,
                    frequency: node.frequency,
                    data: node.data
                });
            }
            return;
        }

        const char = pattern[index];
        
        if (char === '?') {
            // Wildcard: try all children
            for (const childChar of Object.keys(node.children)) {
                this._searchPatternHelper(
                    node.children[childChar],
                    pattern,
                    index + 1,
                    currentWord + childChar,
                    results,
                    limit
                );
            }
        } else {
            // Exact match
            const childNode = node.children[char];
            if (childNode) {
                this._searchPatternHelper(
                    childNode,
                    pattern,
                    index + 1,
                    currentWord + char,
                    results,
                    limit
                );
            }
        }
    }

    /**
     * Get the longest common prefix of all words in the trie
     * @returns {string}
     */
    longestCommonPrefix() {
        let prefix = '';
        let node = this.root;

        while (node) {
            const children = Object.keys(node.children);
            if (children.length !== 1 || node.isEndOfWord) {
                break;
            }
            prefix += children[0];
            node = node.children[children[0]];
        }

        return prefix;
    }

    /**
     * Count words with a given prefix
     * @param {string} prefix - The prefix
     * @returns {number} Number of words with this prefix
     */
    countWithPrefix(prefix) {
        if (!prefix || typeof prefix !== 'string') {
            return 0;
        }

        const node = this._findNode(prefix);
        if (!node) {
            return 0;
        }

        return this._countWords(node);
    }

    /**
     * Count all words from a given node
     * @private
     */
    _countWords(node) {
        let count = node.isEndOfWord ? 1 : 0;
        for (const child of Object.values(node.children)) {
            count += this._countWords(child);
        }
        return count;
    }

    /**
     * Clear the trie
     */
    clear() {
        this.root = new TrieNode();
        this.wordCount = 0;
    }

    /**
     * Get the total number of words in the trie
     * @returns {number}
     */
    size() {
        return this.wordCount;
    }

    /**
     * Check if the trie is empty
     * @returns {boolean}
     */
    isEmpty() {
        return this.wordCount === 0;
    }

    /**
     * Get word data if it exists
     * @param {string} word - The word to look up
     * @returns {*|null} The associated data or null
     */
    getWordData(word) {
        if (!word || typeof word !== 'string') {
            return null;
        }

        const node = this._findNode(word);
        if (node && node.isEndOfWord) {
            return node.data;
        }
        return null;
    }

    /**
     * Update data associated with a word
     * @param {string} word - The word to update
     * @param {*} data - New data to associate
     * @returns {boolean} True if word was found and updated
     */
    updateWordData(word, data) {
        if (!word || typeof word !== 'string') {
            return false;
        }

        const node = this._findNode(word);
        if (node && node.isEndOfWord) {
            node.data = data;
            return true;
        }
        return false;
    }

    /**
     * Export trie to JSON format
     * @returns {Object}
     */
    toJSON() {
        return {
            wordCount: this.wordCount,
            root: this._nodeToJSON(this.root)
        };
    }

    /**
     * Convert node to JSON
     * @private
     */
    _nodeToJSON(node) {
        const json = {
            isEndOfWord: node.isEndOfWord,
            frequency: node.frequency
        };
        
        if (node.data !== null) {
            json.data = node.data;
        }
        
        if (Object.keys(node.children).length > 0) {
            json.children = {};
            for (const [char, child] of Object.entries(node.children)) {
                json.children[char] = this._nodeToJSON(child);
            }
        }
        
        return json;
    }

    /**
     * Import trie from JSON format
     * @param {Object} json - JSON representation of trie
     * @returns {Trie} Returns this for chaining
     */
    fromJSON(json) {
        this.clear();
        this.wordCount = json.wordCount;
        this._nodeFromJSON(this.root, json.root);
        return this;
    }

    /**
     * Convert JSON to node
     * @private
     */
    _nodeFromJSON(node, json) {
        node.isEndOfWord = json.isEndOfWord;
        node.frequency = json.frequency;
        node.data = json.data || null;
        
        if (json.children) {
            for (const [char, childJson] of Object.entries(json.children)) {
                node.children[char] = new TrieNode();
                this._nodeFromJSON(node.children[char], childJson);
            }
        }
    }
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { Trie, TrieNode };
}

if (typeof window !== 'undefined') {
    window.Trie = Trie;
    window.TrieNode = TrieNode;
}